src/Orleans.Runtime/Placement/ResourceOptimizedPlacementDirector.cs

```diff
@@ -0,0 +1,257 @@
+            {
+                relevantSilos.Add(new(silo, stats));
+            }
+        }
+
+        int chooseFrom = (int)Math.Ceiling(Math.Sqrt(relevantSilos.Count));
+        Dictionary<SiloAddress, float> chooseFromSilos = [];
+
+        while (chooseFromSilos.Count < chooseFrom)
+        {
+            int index = Random.Shared.Next(relevantSilos.Count);
+            var pickedSilo = relevantSilos[index];
+
+            relevantSilos.RemoveAt(index);
+
+            float score = CalculateScore(pickedSilo.Value);
+            chooseFromSilos.Add(pickedSilo.Key, score);
+        }
+
+        var orderedByLowestScore = chooseFromSilos.OrderBy(kv => kv.Value);
+
+        // there could be more than 1 silo that has the same score, we pick 1 of them randomly so that we dont continuously pick the first one.
+        var lowestScore = orderedByLowestScore.First().Value;
+        var shortListedSilos = orderedByLowestScore.TakeWhile(p => p.Value == lowestScore).ToList();
+        var winningSilo = shortListedSilos[Random.Shared.Next(shortListedSilos.Count)];
+
+        return winningSilo;
+    }
+
+    bool IsLocalSiloPreferable(IPlacementContext context, SiloAddress[] compatibleSilos, float bestCandidateScore)
+    {
+        if (context.LocalSiloStatus != SiloStatus.Active || !compatibleSilos.Contains(context.LocalSilo))
+        {
+            return false;
+        }
+
+        if (siloStatistics.TryGetValue(context.LocalSilo, out var localStats))
+        {
+            float localScore = CalculateScore(localStats);
+            float scoreDiff = Math.Abs(localScore - bestCandidateScore);
+
+            if (_options.LocalSiloPreferenceMargin >= scoreDiff)
+            {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /// <summary>
+    /// Always returns a value [0-1]
+    /// </summary>
+    /// <returns>
+    /// score = cpu_weight * (cpu_usage / 100) +
+    ///         mem_usage_weight * (mem_usage / physical_mem) +
+    ///         mem_avail_weight * [1 - (mem_avail / physical_mem)]
+    ///         physical_mem_weight * (1 / (1024 * 1024 * physical_mem)
+    /// </returns>
+    /// <remarks>physical_mem is represented in [MB] to keep the result within [0-1] in cases of silos having physical_mem less than [1GB]</remarks>
+    float CalculateScore(ResourceStatistics stats)
+    {
+        float normalizedCpuUsage = stats.CpuUsage.HasValue ? stats.CpuUsage.Value / 100f : 0f;
+
+        if (stats.TotalPhysicalMemory is { } physicalMemory && physicalMemory > 0)
+        {
+            float normalizedMemoryUsage = stats.MemoryUsage.HasValue ? stats.MemoryUsage.Value / physicalMemory : 0f;
+            float normalizedAvailableMemory = 1 - (stats.AvailableMemory.HasValue ? stats.AvailableMemory.Value / physicalMemory : 0f);
+            float normalizedPhysicalMemory = physicalMemoryScalingFactor * physicalMemory;
```

Comment:
```suggestion
            float normalizedPhysicalMemory = PhysicalMemoryScalingFactor * physicalMemory;
```

To match the change I suggested above